plugins/language/plugin.js

```diff
@@ -47,6 +49,27 @@
 				}
 			} );
 
+			// Add missing proper `dir` attribute when <span> element has only `lang` attribute (#5085).
+			editor.on( 'toHtml', function( evt ) {
+				var langElements = evt.data.dataValue.find( function( child ) {
+					return child.name === 'span' &&
+						child.attributes.lang && ( !child.attributes.dir || child.attributes.dir === '' );
+				}, true );
+
+				if ( langElements.length === 0 ) {
+					return;
+				}
+
+				CKEDITOR.tools.array.forEach( langElements, function( element ) {
+					var rtlLanguages = CKEDITOR.tools.object.keys( CKEDITOR.lang.rtl ),
+						isRtlLanguage = CKEDITOR.tools.array.indexOf( rtlLanguages, element.attributes.lang ) !== -1,
+						dirAttribute = isRtlLanguage ? 'rtl' : 'ltr';
+
+					element.attributes.dir = dirAttribute;
+				} );
+
+			}, null, null, 10 );
```

Comment:
I was quite sure that we have some kind of API dedicated for such use cases. And it seems that we, indeed, have one! Although it's undocumented :confused: (see #4839, #5196, or #4585). Yet it can be reverse-engineered through the code of our other plugins, e.g. [`colorbutton`](https://github.com/ckeditor/ckeditor4/blob/3dec7a87364e14c6e40791575c7a537109a5eb05/plugins/colorbutton/plugin.js#L35-L51). Buttons added to the toolbar via the `editor.ui#add()` method can have a `contentTransformations` property that can – as its name suggests – transform content input to the editor. The syntax of transformations is described in detail in [the `editor.filter#addTransformations()` method's docs](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_filter.html#method-addTransformations).

Reply:
I applied `contentTransformations` and it looks much better. Thank you. 😉